### PR TITLE
(CDAP-19134) Tracks and emits all changes in k8s watcher

### PR DESCRIPTION
**Pull request by chtyim**
K8s reference about the semantics

https://kubernetes.io/docs/reference/using-api/_print/#efficient-detection-of-changes

### CODE REVIEW

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java [hunk 1]
```diff
@@ -47,62 +66,62 @@
...
+  private volatile ApiClient apiClient;
+
+  protected AbstractWatcherThread(String threadName, String namespace, String group, String version, String plural) {
+    this(threadName, namespace, group, version, plural, null);
+  }
 
-  protected AbstractWatcherThread(String threadName, String namespace) {
+  @VisibleForTesting
+  AbstractWatcherThread(String threadName, String namespace, String group, String version, String plural,
+                        @Nullable ApiClient apiClient) {
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:90]: where are we calling this constructor (passing non-null apiClient)?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java [hunk 2]
```diff
@@ -184,42 +201,178 @@ public final void run() {
...
+    options = new ListOptions();
+    updateListOptions(options);
+    options.setResourceVersion(resourceVersion);
+    Watchable<DynamicKubernetesObject> watch = wrapWatchableClose(api.watch(namespace, options));
+    Closeables.close(this.watch.getAndSet(watch), true);
+    return watch;
+  }
+
+  private T decodeResource(DynamicKubernetesObject object) throws IOException {
+    return getApiClient().getJSON().getGson().fromJson(object.getRaw(), resourceType);
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:306]: Just for my understanding, why do we need to use the api client to decode the resource? Wouldn't something like GSON.fromJson(object.getRaw(), resourceType) do the same thing?

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java [hunk 3]
```diff
@@ -184,42 +201,178 @@ public final void run() {
...
+      }
+    }
+
+    LOG.debug("Start watching '{}/{}/{}' starting at resource version {}", group, version, plural, resourceVersion);
+
+    // Create the new watch
+    options = new ListOptions();
+    updateListOptions(options);
+    options.setResourceVersion(resourceVersion);
+    Watchable<DynamicKubernetesObject> watch = wrapWatchableClose(api.watch(namespace, options));
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:300]: why do we need this wrapper? Closeables.close won't propagate IOException

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java [hunk 4]
```diff
@@ -47,62 +66,62 @@
...
+  protected AbstractWatcherThread(String threadName, String namespace, String group, String version, String plural) {
+    this(threadName, namespace, group, version, plural, null);
+  }
 
-  protected AbstractWatcherThread(String threadName, String namespace) {
+  @VisibleForTesting
+  AbstractWatcherThread(String threadName, String namespace, String group, String version, String plural,
+                        @Nullable ApiClient apiClient) {
     super(threadName);
+    setDaemon(true);
```
Comment (rmstar) [inline cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:92]: Are we moving the setDaemon() call from child classes to the base class? In some cases, looks like the child class calls setDaemon(true) (ex - in AppResourceWatcherThread https://github.com/cdapio/cdap/pull/14408/files#diff-c493972827be46b0a7ada1ec5226134f35a25de136d2c9282a06b980b3e10d9fL78). But in some other cases it doesn't seem to be called (ex - PodWatcherThread https://github.com/cdapio/cdap/blob/bc69c0a7999a32f304b77855a32db1b04e0d24e0/cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java#L1261)

Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:92]: Yes we are. Just make all these threads daemon threads by default.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java (hunk 1 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:90]: From unit-test, as marked with VisibleForTesting.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java (hunk 3 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:300]: When force closing the Watchable from a different thread, an `IllegalStateException` is throw by the underlying ok.http library, which can be ignored.

File: cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java (hunk 2 above)
Comment (chtyim) [reply cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java:306]: The `Gson` object inside the ApiClient is configured with `TypeAdapter` for decoding k8s object to correct types.

**Review by rmstar — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**